dc/s2n-quic-dc/src/stream/client/rpc.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{event, stream::application::Stream};
+use s2n_quic_core::buffer::{self, writer::Storage};
+use std::{future::Future, io};
+
+pub async fn from_stream<Sub, Req, Res>(
+    stream: Stream<Sub>,
+    mut request: Req,
+    mut response: Res,
+) -> io::Result<Res::Output>
+where
+    Sub: event::Subscriber,
+    Req: Request,
+    Res: Response,
+{
+    let (mut reader, mut writer) = stream.into_split();
+
+    // TODO if the request is large enough, should we spawn a task for it?
+    let writer = async move {
+        while !request.buffer_is_empty() {
+            writer.write_from_fin(&mut request).await?;
+        }
+
+        writer.shutdown()?;
+
+        Ok(())
+    };
+
+    let reader = async move {
+        loop {
+            let storage = response.provide_storage().await?;
+
+            if !storage.has_remaining_capacity() {
+                return Err(io::Error::new(io::ErrorKind::Other, "the provided response buffer failed to provide enough capacity for the peer's response"));
+            }
```

Comment:
nit:

```suggestion
            ensure!(storage.has_remaining_capacity(), Err(io::Error::new(io::ErrorKind::Other, "the provided response buffer failed to provide enough capacity for the peer's response")));
```

Reply:
Good call. I'll fix it in that follow up